File: ui/src/api/loyalty.ts
import { loggedInvoke } from '@/utils/logged-invoke';

/** A loyalty tier defining points thresholds and earn rates. */
export interface LoyaltyTier {
  id: string;
  name: string;
  min_points: number;
  points_per_unit: number;
  earn_multiplier: number;
  colour: string;
  sort_order: number;
  created_at: string;
}

/** A customer's loyalty account with current points balance. */
export interface LoyaltyAccount {
  id: string;
  customer_id: string;
  points: number;
  lifetime_points: number;
  tier_id: string | null;
  updated_at: string;
  created_at: string;
}

/** A points earn or redeem transaction on a loyalty account. */
export interface LoyaltyTransaction {
  id: string;
  account_id: string;
  sale_id: string | null;
  points: number;
  txn_type: string;
  description: string;
  created_at: string;
}

/** A loyalty account with tier info, recent transactions, and next tier progress. */
export interface LoyaltyAccountWithDetails {
  account: LoyaltyAccount;
  tier: LoyaltyTier | null;
  recent_transactions: LoyaltyTransaction[];
  next_tier: LoyaltyTier | null;
  points_to_next_tier: number;
}

/** Result of redeeming loyalty points, with the generated transaction and discount amount. */
export interface RedeemResult {
  transaction: LoyaltyTransaction;
  discount_minor: number;
}

/** Get a loyalty account from the store resolved by the active session. */
export const getLoyaltyAccount = (
  sessionToken: string,
  customerId: string,
): Promise<LoyaltyAccountWithDetails | null> =>
  loggedInvoke<LoyaltyAccountWithDetails | null>('get_loyalty_account_scoped', {
    sessionToken,
    customerId,
  });

/** List loyalty accounts from the store resolved by the active session. */
export const listLoyaltyAccounts = (
  sessionToken: string,
): Promise<LoyaltyAccountWithDetails[]> =>
  loggedInvoke<LoyaltyAccountWithDetails[]>('list_loyalty_accounts_scoped', {
    sessionToken,
  });

/** Earn loyalty points in the store resolved by the active session. */
export const earnLoyaltyPoints = (
  sessionToken: string,
  customerId: string,
  saleId: string,
  totalMinor: number,
): Promise<LoyaltyTransaction> =>
  loggedInvoke<LoyaltyTransaction>('earn_loyalty_points_scoped', {
    sessionToken,
    customerId,
    saleId,
    totalMinor,
  });

/** Redeem loyalty points in the store resolved by the active session. */
export const redeemLoyaltyPoints = (
  sessionToken: string,
  customerId: string,
  points: number,
  saleId: string,
): Promise<RedeemResult> =>
  loggedInvoke<RedeemResult>('redeem_loyalty_points_scoped', {
    sessionToken,
    customerId,
    points,
    saleId,
  });

/** List loyalty tiers from the store resolved by the active session. */
export const listLoyaltyTiers = (sessionToken: string): Promise<LoyaltyTier[]> =>
  loggedInvoke<LoyaltyTier[]>('list_loyalty_tiers_scoped', { sessionToken });

/** Update a loyalty tier in the store resolved by the active session. */
export const updateLoyaltyTier = (
  sessionToken: string,
  tier: LoyaltyTier,
): Promise<LoyaltyTier> =>
  loggedInvoke<LoyaltyTier>('update_loyalty_tier_scoped', { sessionToken, tier });

/** Convert loyalty points into minor currency units in the active store. */
export const getPointsValue = (sessionToken: string, points: number): Promise<number> =>
  loggedInvoke<number>('get_points_value_scoped', { sessionToken, points });

/** Get or create a loyalty account in the store resolved by the active session. */
export const getOrCreateLoyaltyAccount = (
  sessionToken: string,
  customerId: string,
): Promise<LoyaltyAccount> =>
  loggedInvoke<LoyaltyAccount>('get_or_create_loyalty_account_scoped', {
    sessionToken,
    customerId,
  });

File: apps/desktop-client/src/commands/loyalty.rs
use serde::Serialize;
use tauri::State;

use oz_core::db::Store;
use oz_core::loyalty::{
    LoyaltyAccount, LoyaltyAccountWithDetails, LoyaltyTier, LoyaltyTransaction,
};
use oz_core::permissions;

use crate::commands::authz::require_permission_for_user;
use crate::error::AppError;
use crate::state::AppState;

/// Verify a loyalty permission against the global identity database.
///
/// Users and roles are global authentication records; loyalty business data
/// is read from the store-scoped connection after this check succeeds.
async fn require_loyalty_permission(
    state: &AppState,
    user_id: &str,
    permission: &str,
) -> Result<(), AppError> {
    let db = state.db.lock().await;
    let store = Store::new(&db);
    require_permission_for_user(&store, user_id, permission)
}

/// The result of a successful loyalty points redemption.
#[derive(Debug, Serialize)]
pub struct RedeemResult {
    /// The ledger transaction recording the points deduction.
    pub transaction: LoyaltyTransaction,
    /// The calculated discount amount in minor currency units.
    pub discount_minor: i64,
}

/// Retrieves a loyalty account from the store resolved by the active session.
#[tauri::command]
pub async fn get_loyalty_account_scoped(
    session_token: String,
    customer_id: String,
    state: State<'_, AppState>,
) -> Result<Option<LoyaltyAccountWithDetails>, AppError> {
    let (session, conn) = state.resolve_scope(&session_token)?;
    require_loyalty_permission(&state, &session.user_id, permissions::LOYALTY_VIEW).await?;
    let db = conn
        .lock()
        .map_err(|e| AppError::Internal(format!("store db lock: {e}")))?;
    let store = Store::new(&db);
    Ok(store.get_loyalty_account(&customer_id)?)
}

/// Lists loyalty accounts from the store resolved by the active session.
#[tauri::command]
pub async fn list_loyalty_accounts_scoped(
    session_token: String,
    state: State<'_, AppState>,
) -> Result<Vec<LoyaltyAccountWithDetails>, AppError> {
    let (session, conn) = state.resolve_scope(&session_token)?;
    require_loyalty_permission(&state, &session.user_id, permissions::LOYALTY_VIEW).await?;
    let db = conn
        .lock()
        .map_err(|e| AppError::Internal(format!("store db lock: {e}")))?;
    let store = Store::new(&db);
    Ok(store.list_loyalty_accounts()?)
}

/// Awards loyalty points in the store resolved by the active session.
#[tauri::command]
pub async fn earn_loyalty_points_scoped(
    session_token: String,
    customer_id: String,
    sale_id: String,
    total_minor: i64,
    state: State<'_, AppState>,
) -> Result<LoyaltyTransaction, AppError> {
    let (session, conn) = state.resolve_scope(&session_token)?;
    require_loyalty_permission(&state, &session.user_id, permissions::LOYALTY_EARN).await?;
    let db = conn
        .lock()
        .map_err(|e| AppError::Internal(format!("store db lock: {e}")))?;
    let store = Store::new(&db);
    Ok(store.earn_points(&customer_id, &sale_id, total_minor)?)
}

/// Redeems loyalty points in the store resolved by the active session.
#[tauri::command]
pub async fn redeem_loyalty_points_scoped(
    session_token: String,
    customer_id: String,
    points: i64,
    sale_id: String,
    state: State<'_, AppState>,
) -> Result<RedeemResult, AppError> {
    let (session, conn) = state.resolve_scope(&session_token)?;
    require_loyalty_permission(&state, &session.user_id, permissions::LOYALTY_REDEEM).await?;
    let db = conn
        .lock()
        .map_err(|e| AppError::Internal(format!("store db lock: {e}")))?;
    let store = Store::new(&db);
    let (transaction, discount_minor) = store.redeem_points(&customer_id, points, &sale_id)?;
    Ok(RedeemResult {
        transaction,
        discount_minor,
    })
}

/// Lists loyalty tiers from the store resolved by the active session.
#[tauri::command]
pub async fn list_loyalty_tiers_scoped(
    session_token: String,
    state: State<'_, AppState>,
) -> Result<Vec<LoyaltyTier>, AppError> {
    let (session, conn) = state.resolve_scope(&session_token)?;
    require_loyalty_permission(&state, &session.user_id, permissions::LOYALTY_VIEW).await?;
    let db = conn
        .lock()
        .map_err(|e| AppError::Internal(format!("store db lock: {e}")))?;
    let store = Store::new(&db);
    Ok(store.list_tiers()?)
}

/// Updates a loyalty tier in the store resolved by the active session.
#[tauri::command]
pub async fn update_loyalty_tier_scoped(
    session_token: String,
    tier: LoyaltyTier,
    state: State<'_, AppState>,
) -> Result<LoyaltyTier, AppError> {
    let (session, conn) = state.resolve_scope(&session_token)?;
    require_loyalty_permission(&state, &session.user_id, permissions::LOYALTY_MANAGE).await?;
    let db = conn
        .lock()
        .map_err(|e| AppError::Internal(format!("store db lock: {e}")))?;
    let store = Store::new(&db);
    Ok(store.update_tier(
        &tier.id,
        &tier.name,
        tier.min_points,
        tier.points_per_unit,
        tier.earn_multiplier,
        &tier.colour,
    )?)
}

/// Converts loyalty points into minor currency units in the active store.
#[tauri::command]
pub async fn get_points_value_scoped(
    session_token: String,
    points: i64,
    state: State<'_, AppState>,
) -> Result<i64, AppError> {
    let (session, conn) = state.resolve_scope(&session_token)?;
    require_loyalty_permission(&state, &session.user_id, permissions::LOYALTY_VIEW).await?;
    let db = conn
        .lock()
        .map_err(|e| AppError::Internal(format!("store db lock: {e}")))?;
    let store = Store::new(&db);
    Ok(store.get_points_value(points)?)
}

/// Retrieves or creates a loyalty account in the active store.
#[tauri::command]
pub async fn get_or_create_loyalty_account_scoped(
    session_token: String,
    customer_id: String,
    state: State<'_, AppState>,
) -> Result<LoyaltyAccount, AppError> {
    let (session, conn) = state.resolve_scope(&session_token)?;
    require_loyalty_permission(&state, &session.user_id, permissions::LOYALTY_VIEW).await?;
    let db = conn
        .lock()
        .map_err(|e| AppError::Internal(format!("store db lock: {e}")))?;
    let store = Store::new(&db);
    Ok(store.get_or_create_loyalty_account(&customer_id)?)
}

#[cfg(test)] #[path = "loyalty_tests.rs"] mod tests;
